src/Orleans/Lifecycle/LifecycleObservable.cs

```diff
@@ -0,0 +1,116 @@
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans
+{
+    public class LifecycleObservable<TStage> : ILifecycleObservable<TStage>, ILifecycleObserver
+    {
+        private readonly ConcurrentDictionary<object, OrderedObserver> subscribers;
+        private readonly Logger logger;
+        private TStage highStage;
+
+        public LifecycleObservable(Logger logger)
+        {
+            this.logger = logger?.GetLogger(GetType().Name);
+            subscribers = new ConcurrentDictionary<object, OrderedObserver>();
+        }
+
+        public async Task OnStart()
+        {
+            try
+            {
+                foreach (IGrouping<TStage, OrderedObserver> observerGroup in subscribers.Values
+                    .GroupBy(orderedObserver => orderedObserver.Stage)
+                    .OrderBy(group => group.Key))
+                {
+                    highStage = observerGroup.Key;
+                    await Task.WhenAll(observerGroup.Select(orderedObserver => WrapExecution(orderedObserver.Observer.OnStart)));
+                }
+            }
+            catch (Exception ex)
+            {
+                string error = $"Lifecycle start canceled due to errors at stage {this.highStage}";
+                logger?.Error(ErrorCode.LifecycleStartFailure, error, ex);
+                throw new OperationCanceledException(error, ex);
+            }
+        }
+
+        public async Task OnStop()
+        {
+            bool skip = true;
+            foreach (IGrouping<TStage, OrderedObserver> observerGroup in subscribers.Values
+                .GroupBy(orderedObserver => orderedObserver.Stage)
+                .OrderByDescending(group => group.Key))
+            {
+                // skip all until we hit the highest started stage
+                if (skip && highStage.Equals(observerGroup.Key))
+                {
+                    skip = false;
+                }
+                if (skip)
+                {
+                    continue;
+                }
+                highStage = observerGroup.Key;
+                try
+                {
+                    await Task.WhenAll(observerGroup.Select(orderedObserver => WrapExecution(orderedObserver.Observer.OnStop)));
+                }
+                catch (Exception ex)
+                {
+                    logger?.Error(ErrorCode.LifecycleStopFailure, $"Stopping lifecycle encountered an error at stage {this.highStage}.  Continuing to stop.", ex);
+                }
+            }
+        }
+
+        public IDisposable Subscribe(TStage stage, ILifecycleObserver observer)
+        {
+            if (observer == null) throw new ArgumentNullException(nameof(observer));
+
+            var orderedObserver = new OrderedObserver(stage, observer);
+            subscribers.TryAdd(orderedObserver, orderedObserver);
+            return new Disposable(() => Remove(orderedObserver));
+        }
+
+        private void Remove(object key)
```

Comment:
>This method wraps TryRemove and this way swallows the result of the removal, which can fail.

TryRemove would only fail if the key was not found.  Why do we care if the remove failed or not?

>"misusing" IDisposable

How are we misusing IDisposable?

>if the registered Observer is an IDisposable then we should properly dispose it

The observer is not disposable, the subscription is.  Disposing the disposable returned by the subscribe call removes the subscription.